control/control.go

```diff
@@ -886,22 +921,79 @@ func (p *pluginControl) GetMetricVersions(ns core.Namespace) ([]core.CatalogedMe
 	return rmts, nil
 }
 
-func (p *pluginControl) MetricExists(mns core.Namespace, ver int) bool {
-	_, err := p.metricCatalog.Get(mns, ver)
+func (p *pluginControl) MetricsExists(mns core.Namespace, ver int) bool {
+	_, err := p.metricCatalog.GetMetric(mns, ver)
 	if err == nil {
 		return true
 	}
 	return false
 }
 
+// gatherMetricTypesGroupedByPlugin gathers matched metrics types for requested metrics and groups them by plugins to call
+func (p *pluginControl) gatherMetricTypesGroupedByPlugin(rmts []core.RequestedMetric, cdt *cdata.ConfigDataTree) (map[string]metricTypes, []error) {
+	pmts := make(map[string]metricTypes)
+	var errs []error
+
+	if cdt == nil {
+		// initialize config tree if nil is provided
+		cdt = cdata.NewTree()
+	}
+
+	for _, rmt := range rmts {
+		// for each requested metric get matched metrics types on appropriate version
+		// (if version < 1 choose the latest)
+		mts, err := p.metricCatalog.GetMatchedMetricTypes(rmt.Namespace(), rmt.Version())
+
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		for _, mt := range mts {
+			lp := mt.Plugin
+			if lp == nil {
+				errs = append(errs, errorMetricNotFound(mt.Namespace().Strings(), mt.Version()))
+				continue
+			}
+
+			// get config
+			config := cdt.Get(mt.Namespace().Strings())
+
+			if config == nil {
+				config = cdata.NewNode()
+			}
+			returnedmt := plugin.MetricType{
+				Namespace_:          mt.Namespace(),
+				LastAdvertisedTime_: mt.LastAdvertisedTime(),
+				Version_:            mt.Version(),
+				Tags_:               mt.Tags(),
+				Config_:             config,
+				Unit_:               mt.Unit(),
+			}
+
+			// group matched metrics by plugin
+			pluginKey := lp.Key()
+			pmt, _ := pmts[pluginKey]
+			pmt.plugin = lp
```

Comment:
That fragment of code comes from snap master (https://github.com/intelsdi-x/snap/blob/master/control/control.go#L1080), I just renamed this function from **groupMetricTypesByPlugin** to **gatherMetricTypesGroupedByPlugin** to be already more accurate because of retrieving metricTypes from the catalog based on incoming requested metrics namespaces.  But returning to your question - if in "pmts" there is no items (first iteration) or there is no key="pluginKey", **pmt** will be an empty structure. So as the next step pmt.plugin is set (new or just re-set), and **returnedmt** will be appended to **pmt.metricTypes**, and item under **pmts[pluginKey]** will be added/updated. Only thing what can be changed is do not set pmt.lp again if a map contains pluginKey - but it is still not related to feature of specifying instance of dynamic metric. It should be treat  perhaps as improving the snap framework. What do You think?